### PR TITLE
Fix binlog end log pos over lost data

### PR DESCRIPTION
**Pull request by shaohk**
Related issue: https://github.com/github/gh-ost/issues/1366

### Description

This PR [Fix binlog end log pos over lost data]

> In case this PR introduced Go code changes:

- [ ] contributed code is using same conventions as original code
- [ ] `script/cibuild` returns with no formatting errors, build errors or unit test errors.

### CODE REVIEW

**Comment by dikang123** [comment]
when will this PR be merged?

**Comment by timvaillancourt** [comment]
@shaohk thanks a lot for this PR, this sounds like a difficult problem to debug!

Can you think of any ways that this check can be tested in go unit tests? I wonder if a test could:
1. Test that a normal `*replication.RowsEvent` event will be accepted
2. Test that an oversized `*replication.RowsEvent` event it not accepted

Luckily, `handleRowEvents` doesn't need a real MySQL server as it pushes events to `entriesChannel chan<- *BinlogEntry`, which should make this relatively simple to test

**Review by timvaillancourt — CHANGES_REQUESTED** [review]
Looks good! Added one comment requesting unit tests 🙇

**Comment by shaohk** [comment]
> @shaohk thanks a lot for this PR, this sounds like a difficult problem to debug!
> 
> Can you think of any ways that this check can be tested in go unit tests? I wonder if a test could:
> 
> 1. Test that a normal `*replication.RowsEvent` event will be accepted
> 2. Test that an oversized `*replication.RowsEvent` event it not accepted
> 
> Luckily, `handleRowEvents` doesn't need a real MySQL server as it pushes events to `entriesChannel chan<- *BinlogEntry`, which should make this relatively simple to test

@timvaillancourt Thank you for reviewing this PR. I have added unit tests as requested. Could you please take a look again? Thank you for your hard work.

File: go/mysql/binlog.go [hunk 1]
```diff
@@ -74,3 +75,23 @@ func (this *BinlogCoordinates) SmallerThanOrEquals(other *BinlogCoordinates) boo
 	}
 	return this.LogFile == other.LogFile && this.LogPos == other.LogPos
 }
+
+// IsLogPosOverflowBeyond4Bytes returns true if the coordinate endpos is overflow beyond 4 bytes.
```
Comment (timvaillancourt) [inline go/mysql/binlog.go:79]: @shaohk last request, can we add a reference to why a value over 4 bytes is an overflow?

Comment (shaohk) [reply go/mysql/binlog.go:79]: OK, I have added. Please take a look again, Thks.

**Comment by meiji163** [comment]
@shaohk Can you take a look at the failing tests and linter?

**Comment by shaohk** [comment]
> @shaohk Can you take a look at the failing tests and linter?

@meiji163 
OK, Please take a look again, Thks.

**Review by meiji163 — APPROVED** [review]

**Comment by meiji163** [comment]
LGTM, can you take another look @timvaillancourt?

**Comment by timvaillancourt** [comment]
> LGTM, can you take another look @timvaillancourt?

@meiji163 / @shaohk: I think erroring here is the easiest next step, so LGTM 👍

But longer term, what do we "want" `gh-ost` to do here? Or to put it differently, what does the MySQL replica do in this scenario? I'm assuming it doesn't error right? 🤔

**Review by timvaillancourt — APPROVED** [review]
Approved with longer-term question

**Comment by meiji163** [comment]
I'm not sure how MySQL handles it
I found this bug report from last year https://bugs.mysql.com/bug.php?id=112189

**Comment by shaohk** [comment]
> > LGTM, can you take another look @timvaillancourt?
> 
> @meiji163 / @shaohk: I think erroring here is the easiest next step, so LGTM 👍
> 
> But longer term, what do we "want" `gh-ost` to do here? Or to put it differently, what does the MySQL replica do in this scenario? I'm assuming it doesn't error right? 🤔

https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#sysvar_max_binlog_cache_size.

`When [gtid_mode](https://dev.mysql.com/doc/refman/8.0/en/replication-options-gtids.html#sysvar_gtid_mode) is not ON, the maximum recommended value is 4GB, due to the fact that, in this case, MySQL cannot work with binary log positions greater than 4GB; when gtid_mode is ON, this limitation does not apply, and the server can work with binary log positions of arbitrary size.`

So I believe that MySQL master-slave replication is based on the GTID mode, so there is no problem with MySQL master-slave replication. The gh-ost tool for synchronizing binlogs uses a non-GTID mode, which is why this issue arises.

**Pull request merged**